plugins/src/main/java/com/google/fhir/proxy/plugin/OpenSRPSyncAccessDecision.java

```diff
@@ -0,0 +1,168 @@
+		this.applicationId = applicationId;
+		this.accessGranted = accessGranted;
+		this.careTeamIds = careTeamIds;
+		this.locationIds = locationIds;
+		this.organizationIds = organizationIds;
+		this.syncStrategy = syncStrategy;
+	}
+
+	@Override
+	public boolean canAccess() {
+		return accessGranted;
+	}
+
+	@Override
+	public void preProcess(ServletRequestDetails servletRequestDetails) {
+		// TODO: Disable access for a user who adds tags to organisations, locations or care teams that they do not have access to
+		//  This does not bar access to anyone who uses their own sync tags to circumvent
+		//  the filter. The aim of this feature based on scoping was to pre-filter the data for the user
+		if (isSyncUrl(servletRequestDetails)) {
+			// This prevents access to a user who has no location/organisation/team assigned to them
+			if (locationIds.size() == 0 && careTeamIds.size() == 0 && organizationIds.size() == 0) {
+				locationIds.add(
+						"CR1bAeGgaYqIpsNkG0iidfE5WVb5BJV1yltmL4YFp3o6mxj3iJPhKh4k9ROhlyZveFC8298lYzft8SIy8yMNLl5GVWQXNRr1sSeBkP2McfFZjbMYyrxlNFOJgqvtccDKKYSwBiLHq2By5tRupHcmpIIghV7Hp39KgF4iBDNqIGMKhgOIieQwt5BRih5FgnwdHrdlK9ix");
+			}
+			addSyncTags(servletRequestDetails, getSyncTags(locationIds, careTeamIds, organizationIds));
+		}
+	}
+
+	private void addSyncTags(ServletRequestDetails servletRequestDetails, Pair<String, Map<String, String[]>> syncTags) {
+
+		List<String> params = new ArrayList<>();
+
+		for (Map.Entry<String, String[]> entry : syncTags.getValue().entrySet()) {
+			String tagName = entry.getKey();
+			for (String tagValue : entry.getValue()) {
+				StringBuilder sb = new StringBuilder(tagName.length() + tagValue.length() + 2);
+				sb.append(tagName);
+				sb.append("|");
+				sb.append(tagValue);
+				params.add(sb.toString());
+			}
+		}
+
+		String[] prevTagFilters = servletRequestDetails.getParameters().get(SEARCH_PARAM_TAG);
+		if (prevTagFilters != null && prevTagFilters.length > 1) {
+			Collections.addAll(params, prevTagFilters);
+		}
+
+		servletRequestDetails.addParameter(SEARCH_PARAM_TAG, params.toArray(new String[0]));
+	}
+
+	@Override
+	public String postProcess(HttpResponse response) throws IOException {
+		return null;
+	}
+
+	private Pair<String, Map<String, String[]>> getSyncTags(List<String> locationIds, List<String> careTeamIds,
+			List<String> organizationIds) {
+		StringBuilder sb = new StringBuilder();
+		Map<String, String[]> map = new HashMap<>();
+
+		sb.append(SEARCH_PARAM_TAG);
+		sb.append("=");
+		addTags(LOCATION_TAG_URL, locationIds, map, sb);
+		addTags(ORGANISATION_TAG_URL, organizationIds, map, sb);
+		addTags(CARE_TEAM_TAG_URL, careTeamIds, map, sb);
+
+		return new ImmutablePair<>(sb.toString(), map);
+	}
+
+	private void addTags(String tagUrl, List<String> values, Map<String, String[]> map, StringBuilder sb) {
+		int len = values.size();
+		if (len > 0) {
+			if (sb.length() != (SEARCH_PARAM_TAG + "=").length()) {
```

Comment:
Can we move these string literals to the constant file?

Reply:
Definitely, can we create another constants file inside the plugins instead of the server?

Reply:
Yes

Reply:
Cool, can we create an issue for this and make these changes even in `permission-checker-and-data-access` branch

Reply:
I noticed that we also have some other OpenSRP specific constants such as `SYNC_STRATEGY` in the `ProxyConstants` file

Reply:
Yes, I think we can create a separate issue for this and cater it in the later part once we move plugins to the OpenSRP specific repo (Extensions module). @dubdabasoduba  @ekigamba